### PR TITLE
feat: virtualize pools for the AMM.

### LINKED ISSUE #4344: Use Virtual Objects in AMM
## What is the Problem Being Solved?

For MainNet 1, we intend to support durable objects for restart and upgrade.

## Description of the Design

Collect a catalogue of all the persistent data types in the AMM, and convert them to Virtual Objects.

## Security Considerations

This is about durability. It shouldn't have broader security implications.

## Test Plan

Presumably extensive.

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
closes: #4344

## Description

split out addPool() to a new file.
separate facets for pool, vpoolOuter, vpoolInner, and helpers

helpers is used for behavior (addliquidityActual) that was passed as a function
vPoolOuter is the singlePool that is polymorphic with doublePool.

### Security Considerations

Just a refactoring for virtualization. Shouldn't have security consequences.

### Documentation Considerations

No visible impact to clients.

### Testing Considerations

None.

### CODE REVIEW

File: packages/SwingSet/docs/virtual-objects.md [hunk 1]
```diff
@@ -51,7 +51,16 @@ where `context` describes the invocation context of the method and `...args` are
 
 The `options` parameter is optional. It provides additional parameters to characterize the VDO.  Currently there is only one supported option, `finish`, though we anticipate more options may be added in future versions of the API.
 
-The `finish` option is a function that, if present, will be called at the end of instance initialization.  It will be invoked after the VDO is created but before it is returned from the maker function.  `finish` is passed one parameters, a `context` object identical to that passed to method behaviors.  The `finish` function can modify the object's state at a time when the object's identity is known (or its facets' identies are known), and thus can be used in cases where a validly initialized instance requires it to participate in some kind of cyclical object graph with other VDOs.  It can also be used, for example, to register the object with outside tracking data structures, or do whatever other post-creation setup is needed for the object to do its job.
+The `finish` option is a function that, if present, will be called at the end of
```
Comment (turadg) [inline packages/SwingSet/docs/virtual-objects.md:54]: is this is line wrapping? `feat: virtualize pools for the AMM` could be a confusing git blame

File: packages/run-protocol/src/vpool-xyk-amm/doublePool.js [hunk 2]
```diff
@@ -179,20 +170,9 @@ export const makeDoublePool = (
...
-    return publicPrices(getPriceForOutput(amountIn, amountOut));
-  };
-
-  const externalFacet = Far('double pool', {
-    getInputPrice,
-    getOutputPrice,
-  });
-
-  const internalFacet = Far('single pool', {
+  return Far('single pool', {
```
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/doublePool.js]: typo? this is what's returned by `makeDoublePool`.

File: packages/run-protocol/src/vpool-xyk-amm/pool.js [hunk 3]
```diff
@@ -1,301 +1,328 @@
...
  * @param {ERef<Timer>} timer
  * @param {IssuerKit} quoteIssuerKit
- * @param {() => bigint} getProtocolFeeBP - retrieve governed protocol fee value
- * @param {() => bigint} getPoolFeeBP - retrieve governed pool fee value
- * @param {ZCFSeat} protocolSeat
+ * @param {import('./multipoolMarketMaker.js').AMMParamGetters} paramAccessor retrieve governed params
+ * @param {ZCFSeat} protocolSeat seat that holds collected fees
  */
-export const makeAddPool = (
+export const makePoolMaker = (
```
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/pool.js]: "make … maker" is a smell to me. Fortunately we have a more specific thing we're making: a "kind" is an instance of the thing a `maker` returned by `defineKind` makes. Cutting to the chase:
```suggestion
export const definePoolKind = (
```

And then in `addPool.js`,
```
  const makePool = definePoolKind(
```

File: packages/run-protocol/src/vpool-xyk-amm/singlePool.js [hunk 4]
```diff
@@ -7,89 +7,72 @@ import {
...
+const getCentralBrand = pool => pool.getCentralAmount().brand;
 
-  const allocateGainsAndLosses = (seat, prices) => {
-    const poolSeat = pool.getPoolSeat();
+const getPools = pool => ({
+  Central: pool.getCentralAmount(),
+  Secondary: pool.getSecondaryAmount(),
+});
+
+export const singlePool = Far('SinglePool', {
```
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/singlePool.js]: why `Far` this? It's a behavior object. I think this could be,
```suggestion
export const singlePoolBehavior = {
```

File: packages/run-protocol/src/vpool-xyk-amm/pool.js [hunk 5]
```diff
@@ -1,301 +1,328 @@
...
+ * @typedef {Readonly<{
+ * liquidityBrand: Brand<'nat'>,
+ * secondaryBrand: Brand<'nat'>,
+ * centralBrand: Brand<'nat'>,
+ * liquidityZcfMint: ZCFMint,
+ * liquidityIssuer: Issuer<any>,
+ * toCentralPriceAuthority: PriceAuthority,
+ * fromCentralPriceAuthority: PriceAuthority,
+ * quoteIssuerKit: IssuerKit,
+ * zcf: ZCF,
```
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/pool.js:37]:
```suggestion
 * protocolSeat: ZCFSeat,
 * zcf: ZCF,
```
Discovered when I types:
```
  /**
   * @param {import('./pool').MethodContext} context
   * @param {ZCFSeat} seat
   * @param {*} prices
   */
  allocateGainsAndLosses: (context, seat, prices) => {
```

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    // before proceeding), so we can do the work now.
+    await zcf.saveIssuer(secondaryIssuer, keyword);
+    const liquidityZCFMint = await zcf.makeZCFMint(
+      liquidityKeyword,
+      AssetKind.NAT,
+      harden({ decimalPlaces: 6 }),
+    );
+    const { zcfSeat: poolSeat } = zcf.makeEmptySeatKit();
+    const pool = makePool(liquidityZCFMint, poolSeat, secondaryBrand);
+    // @ts-ignore xxx fix types
```
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/addPool.js:77]: this is because `DoublePoolSwapResult` requires many more methods than `SwapResults` provides. Out of scope for this PR but something to address in how the pools work.

If it weren't for time pressure I'd request the pool refactoring be separated from this PR so this can be about the virtualization per se.

File: packages/run-protocol/src/vpool-xyk-amm/pool.js [hunk 7]
```diff
@@ -1,301 +1,328 @@
...
+ * poolSeat: ZCFSeat,
+ * liqTokenSupply: bigint,
+ * }} MutableState
+ *
+ * @typedef {{
+ *   state: ImmutableState & MutableState,
+ *   facets: {
+ *     helper: import('@agoric/vat-data/src/types').KindFacet<typeof helperBehavior>,
+ *     pool: import('@agoric/vat-data/src/types').KindFacet<any>, // FIXME typeof poolBehavior
+ *     singlePool: import('@agoric/vat-data/src/types').KindFacet<VirtualPool>,
```
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/pool.js]: `KindFacet` makes a facet from a behavior. `VirtualPool` is already a facet so,
```suggestion
 *     singlePool: VirtualPool,
```
or,
```
 *     singlePool: import('@agoric/vat-data/src/types').KindFacet<typeof singlePoolBehavior>,
```

File: packages/run-protocol/src/vpool-xyk-amm/pool.js [hunk 8]
```diff
@@ -1,301 +1,328 @@
...
+ * notifier: Notifier<any>,
+ * poolSeat: ZCFSeat,
+ * liqTokenSupply: bigint,
+ * }} MutableState
+ *
+ * @typedef {{
+ *   state: ImmutableState & MutableState,
+ *   facets: {
+ *     helper: import('@agoric/vat-data/src/types').KindFacet<typeof helperBehavior>,
+ *     pool: import('@agoric/vat-data/src/types').KindFacet<any>, // FIXME typeof poolBehavior
```
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/pool.js]:
```suggestion
 *     pool: import('@agoric/vat-data/src/types').KindFacet<typeof poolBehavior>,
```

File: packages/run-protocol/src/vpool-xyk-amm/pool.js [hunk 9]
```diff
@@ -1,301 +1,328 @@
...
+const poolBehavior = {
+  getLiquiditySupply: ({ state: { liqTokenSupply } }) => liqTokenSupply,
+  getLiquidityIssuer: ({ state: { liquidityIssuer } }) => liquidityIssuer,
+  getPoolSeat: ({ state: { poolSeat } }) => poolSeat,
+  getCentralAmount: ({ state: { poolSeat, centralBrand } }) =>
+    poolSeat.getAmountAllocated('Central', centralBrand),
+  getSecondaryAmount: ({ state }) =>
+    state.poolSeat.getAmountAllocated('Secondary', state.secondaryBrand),
+
+  addLiquidity: ({ state, facets: { helper, pool } }, zcfSeat) => {
```
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/pool.js]: Not gonna block on this but it would help future refactoring to have types here. I tried it and there's a gotcha with circular references, so that the `facets` can't be destructured. There's also a circular definition of the return value so that needs to be specified,
```suggestion
  /**
   * @param {MethodContext} context
   * @param {ZCFSeat} zcfSeat
   * @returns {string}
   */
  addLiquidity: ({ state, facets }, zcfSeat) => {
    const { helper, pool } = facets;
```

File: packages/run-protocol/src/vpool-xyk-amm/pool.js [hunk 10]
```diff
@@ -1,301 +1,328 @@
...
-  };
+  const facets = harden({
+    helper: helperBehavior,
+    pool: poolBehavior,
+    singlePool,
+  });
 
-  return addPool;
+  // @ts-ignore unhappy about finish's type
+  return defineKindMulti('pool', poolInit, facets, { finish });
```
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/pool.js:334]: the function that wraps this seems to be used for currying, but since there are parameters it means if they change so does the kind. 

The only one that looks like it would qualify the pool kind is the central brand. Consider:

```suggestion
  const tag = `pool for ${centralBrand.getAllegedName()}`;
  return defineKindMulti(tag, poolInit, facets, { finish });
```

**Review by turadg — COMMENTED** [review]
I've been working on a review and I see some commits so I'll send it before there are more :)

File: packages/SwingSet/docs/virtual-objects.md (hunk 1 above)
Comment (Chris-Hibbert) [reply packages/SwingSet/docs/virtual-objects.md:54]: If you insist, I can put the one-very-long-line back, but I consider it a mis-feature. It makes it hard to read in the source, hard to edit, and hard to review changes.

I changed "`finish` is passed one parameters" to "`finish` is passed one parameter", because it took me  a while to decide that the typo was "parameters" rather than "one". I thought there was a missing parameter description for a while.

File: packages/run-protocol/src/vpool-xyk-amm/doublePool.js (hunk 2 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/doublePool.js]: fixed.

File: packages/run-protocol/src/vpool-xyk-amm/singlePool.js (hunk 4 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/singlePool.js]: fixed.

File: packages/run-protocol/src/vpool-xyk-amm/pool.js (hunk 9 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js]: done

File: packages/run-protocol/src/vpool-xyk-amm/pool.js (hunk 7 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js]: done

File: packages/run-protocol/src/vpool-xyk-amm/pool.js (hunk 3 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js]: I'm used to seeing `makeFooMaker`. I've seen `FooFactoryFactory`.  can you summarize what bad practice it indicates?

File: packages/run-protocol/src/vpool-xyk-amm/pool.js (hunk 10 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js:334]: At first I thought central was wrong, but now I agree that it's common across the pools.  Unfortunately, the central brand is remote, so this would require an `await`. I don't think it's justified.

File: packages/run-protocol/src/vpool-xyk-amm/pool.js (hunk 5 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js:37]: added

File: packages/run-protocol/src/vpool-xyk-amm/pool.js (hunk 8 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js]: I started to do that, but discovered that it's exactly `XYKPool`, so that's what I changed to.

File: packages/run-protocol/src/vpool-xyk-amm/pool.js (hunk 3 above)
Comment (turadg) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js]: Cognitive burden distinguishing the two things named same (maker, factory). But that's a weak argument.

The better reason to use `definePoolKind` is because it's returning what a `defineKind` returns.

**Review by turadg — APPROVED** [review]

File: packages/run-protocol/src/vpool-xyk-amm/pool.js (hunk 3 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js]: > The better reason to use definePoolKind is because it's returning what a defineKind returns.

I don't disagree that `definePoolKind` is better here. I was looking for a reason to be suspicious of `makeFooMaker` in general.

Comment (turadg) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js]: > I was looking for a reason to be suspicious of makeFooMaker in general.

My error in casting suspicion on it by using the term "smell". I didn't mean a "code smell" (i.e. "a surface indication that usually corresponds to a deeper problem in the system"). I just meant "stinky". :)
